File: RadixCyclicNN/radixnet/activation.py
"""Parametric sine activation.

Every node of the graph owns an activation ``f(x) = a * sin(b * (x - h)) + k``.
With the default parameters this is *exactly* the author's activation
``-1 * sin(x / 3.0)``.  All four parameters are learnable per node, which is
what "update the activation function instead of fighting the vanishing
gradient" means in this design: learning acts on the activation itself.

The hot loops in :mod:`radixnet.backend` inline these formulas (no function
calls per element); the functions here are the reference implementation used
by the graph, the tests and the documentation.
"""

from __future__ import annotations

import math

DEFAULT_A = -1.0
DEFAULT_B = 1.0 / 3.0
DEFAULT_H = 0.0
DEFAULT_K = 0.0


def sine_activation(
    x: float,
    a: float = DEFAULT_A,
    b: float = DEFAULT_B,
    h: float = DEFAULT_H,
    k: float = DEFAULT_K,
) -> float:
    """Return ``a * sin(b * (x - h)) + k`` (default: ``-sin(x / 3)``)."""
    return a * math.sin(b * (x - h)) + k


def sine_derivative(
    x: float,
    a: float = DEFAULT_A,
    b: float = DEFAULT_B,
    h: float = DEFAULT_H,
    k: float = DEFAULT_K,
) -> float:
    """Return ``df/dx = a * b * cos(b * (x - h))``."""
    return a * b * math.cos(b * (x - h))


def sine_partials(
    x: float,
    a: float = DEFAULT_A,
    b: float = DEFAULT_B,
    h: float = DEFAULT_H,
    k: float = DEFAULT_K,
) -> tuple[float, float, float, float, float, float]:
    """Return ``(f, df/dx, df/da, df/db, df/dh, df/dk)`` at ``x``.

    With ``u = b * (x - h)``::

        f     = a * sin(u) + k
        df/dx = a * b * cos(u)
        df/da = sin(u)
        df/db = a * (x - h) * cos(u)
        df/dh = -a * b * cos(u)
        df/dk = 1
    """
    d = x - h
    u = b * d
    su = math.sin(u)
    cu = math.cos(u)
    abc = a * b * cu
    return (a * su + k, abc, su, a * d * cu, -abc, 1.0)


class SineActivation:
    """Convenience object wrapping one parameter set (not used in hot loops)."""

    __slots__ = ("a", "b", "h", "k")

    def __init__(
        self,
        a: float = DEFAULT_A,
        b: float = DEFAULT_B,
        h: float = DEFAULT_H,
        k: float = DEFAULT_K,
    ) -> None:
        self.a = float(a)
        self.b = float(b)
        self.h = float(h)
        self.k = float(k)

    def __call__(self, x: float) -> float:
        """Evaluate the activation at ``x``."""
        return sine_activation(x, self.a, self.b, self.h, self.k)

    def derivative(self, x: float) -> float:
        """Evaluate ``df/dx`` at ``x``."""
        return sine_derivative(x, self.a, self.b, self.h, self.k)

    def partials(self, x: float) -> tuple[float, float, float, float, float, float]:
        """Return ``(f, df/dx, df/da, df/db, df/dh, df/dk)`` at ``x``."""
        return sine_partials(x, self.a, self.b, self.h, self.k)

    def inverted(self) -> "SineActivation":
        """Return a copy with the amplitude sign flipped (``a -> -a``)."""
        return SineActivation(-self.a, self.b, self.h, self.k)

    def to_dict(self) -> dict:
        """JSON-serialisable representation."""
        return {"a": self.a, "b": self.b, "h": self.h, "k": self.k}

    @classmethod
    def from_dict(cls, d: dict) -> "SineActivation":
        """Inverse of :meth:`to_dict`; missing keys fall back to the defaults."""
        return cls(
            d.get("a", DEFAULT_A),
            d.get("b", DEFAULT_B),
            d.get("h", DEFAULT_H),
            d.get("k", DEFAULT_K),
        )

    def __repr__(self) -> str:
        return f"SineActivation(a={self.a!r}, b={self.b!r}, h={self.h!r}, k={self.k!r})"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, SineActivation):
            return NotImplemented
        return (self.a, self.b, self.h, self.k) == (other.a, other.b, other.h, other.k)

    def __hash__(self) -> int:
        return hash((self.a, self.b, self.h, self.k))


def edge_signal(w: float, fp: float, fc: float) -> float:
    """Signal carried by edge ``p -> c``: weight times ``f_p`` times ``f_c``.

    This is the "activation of the child times the activation of the parent"
    rule; the backend uses the same product for every score.
    """
    return w * fp * fc

File: RadixCyclicNN/rust/src/activation.rs
//! The parametric sine activation of the radix model.
//!
//! Every node of a radix graph owns an activation `f(x) = a * sin(b * (x - h)) + k`;
//! with the default parameters it is exactly the author's `-sin(x / 3)`.  All
//! four parameters are learnable per node, which is what "update the activation
//! function instead of fighting the vanishing gradient" means in this design.
//!
//! A port of `radixnet/activation.py`.  The formulas are written in Python's
//! operation order on purpose - `a * sin(b * (z - h)) + k`, never a fused or
//! reassociated form - because the edge scores built from them are compared
//! with Python's to the bit.

/// The default amplitude: negative, so a fresh node reads `-sin(x / 3)`.
pub const DEFAULT_A: f64 = -1.0;
/// The default frequency.
pub const DEFAULT_B: f64 = 1.0 / 3.0;
/// The default phase shift.
pub const DEFAULT_H: f64 = 0.0;
/// The default offset.
pub const DEFAULT_K: f64 = 0.0;

/// `a * sin(b * (x - h)) + k`.
#[inline]
pub fn sine(x: f64, a: f64, b: f64, h: f64, k: f64) -> f64 {
    a * (b * (x - h)).sin() + k
}

/// `df/dx = a * b * cos(b * (x - h))`.
pub fn sine_derivative(x: f64, a: f64, b: f64, h: f64) -> f64 {
    a * b * (b * (x - h)).cos()
}

/// `(f, df/dx, df/da, df/db, df/dh)` at `x` (`df/dk` is 1).
///
/// With `u = b * (x - h)`: `f = a sin u + k`, `df/dx = a b cos u`,
/// `df/da = sin u`, `df/db = a (x - h) cos u`, `df/dh = -a b cos u` - the
/// tuple the backend caches per node and batch.
#[inline]
pub fn partials(x: f64, a: f64, b: f64, h: f64, k: f64) -> [f64; 5] {
    let d = x - h;
    let u = b * d;
    let su = u.sin();
    let cu = u.cos();
    let abc = a * b * cu;
    [a * su + k, abc, su, a * d * cu, -abc]
}

/// The signal edge `p -> c` carries: its weight times both activations.
#[inline]
pub fn edge_signal(w: f64, fp: f64, fc: f64) -> f64 {
    w * fp * fc
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_default_is_minus_sin_of_a_third() {
        for x in [-4.0f64, -1.5, 0.0, 0.7, 3.0] {
            // `b = 1/3` multiplies where `x / 3` divides, so the last bit may differ
            assert!((sine(x, DEFAULT_A, DEFAULT_B, DEFAULT_H, DEFAULT_K) + (x / 3.0).sin()).abs() < 1e-15);
        }
    }

    #[test]
    fn the_partials_match_finite_differences() {
        let (x, a, b, h, k) = (0.8, -1.3, 0.4, 0.2, 0.1);
        let p = partials(x, a, b, h, k);
        assert_eq!(p[0], sine(x, a, b, h, k));
        let eps = 1e-6;
        let fd = |f: &dyn Fn(f64) -> f64| (f(eps) - f(-eps)) / (2.0 * eps);
        assert!((p[1] - fd(&|d| sine(x + d, a, b, h, k))).abs() < 1e-7);
        assert!((p[2] - fd(&|d| sine(x, a + d, b, h, k))).abs() < 1e-7);
        assert!((p[3] - fd(&|d| sine(x, a, b + d, h, k))).abs() < 1e-7);
        assert!((p[4] - fd(&|d| sine(x, a, b, h + d, k))).abs() < 1e-7);
        assert_eq!(p[1], sine_derivative(x, a, b, h));
    }

    #[test]
    fn negating_amplitude_and_offset_negates_the_unit() {
        let (x, a, b, h, k) = (1.1, -0.9, 0.3, 0.05, 0.2);
        assert_eq!(sine(x, -a, b, h, -k), -sine(x, a, b, h, k));
        assert_eq!(edge_signal(2.0, 0.5, -0.25), -0.25);
    }
}
